Support Callbacks on Asynchronous RPC

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24118 Support Callbacks on Asynchronous RPC**
* #23968 Use c10::ThreadPool to send messages

RPC applications can now register callbacks to the FutureMessage
object returned by the asynchronous RPC API. Multiple callbacks can be
chained together by calling `FutureMessage.then(cb1).then(cb2)`. Each
callback should take one parameter, which is the reference to the
`FutureMessage` object. It is up to the callback whether it needs to
retrieve the return value or disregard.

Differential Revision: [D16743652](https://our.internmc.facebook.com/intern/diff/D16743652/)

mrshenli committed:
Support Callbacks on Asynchronous RPC

RPC applications can now register callbacks to the FutureMessage
object returned by the asynchronous RPC API. Multiple callbacks can be
chained together by calling `FutureMessage.then(cb1).then(cb2)`. Each
callback should take one parameter, which is the reference to the
`FutureMessage` object. It is up to the callback whether it needs to
retrieve the return value or disregard.

Differential Revision: [D16743652](https://our.internmc.facebook.com/intern/diff/D16743652/)

mrshenli committed:
Update on "Support Callbacks on Asynchronous RPC"


RPC applications can now register callbacks to the FutureMessage
object returned by the asynchronous RPC API. Multiple callbacks can be
chained together by calling `FutureMessage.then(cb1).then(cb2)`. Each
callback should take one parameter, which is the reference to the
`FutureMessage` object. It is up to the callback whether it needs to
retrieve the return value or disregard.

Differential Revision: [D16743652](https://our.internmc.facebook.com/intern/diff/D16743652/)

pritamdamania87 commented:
@mrshenli Could you add more details about the use case for this? Also, if there is an issue which explains the need for this change in more detail, could you link it in this PR?

mrshenli commented:
@pritamdamania87 Sure, good point, I will add some comments later (I rebased to #23968, which is on top of #23569, and I cannot export before #23569 is landed ).

In short, this is required for implementing `RRef` (as pointed out by @xush6528). One example would be making `torch.remote` calls with implicit `RRef[T]->T` conversion. We would like to kick of the function execution when `RRef.to_here()` finishes instead of blocking and waiting.

(This is NON-GOAL for this PR) More generally, this is also necessary for implementing asynchronous Python UDFs. Consider the following case:

```python
def udf():
  fut = torch.distributed.rpc("worker2", torch.add, torch.ones(2, 2), 1)
  # I don't want to wait, I would like to yield and resume computation 
  # when the async call is done
  return 2 * fut.wait()
```

With the current API, UDF cannot yield, as a result, on the callee `RpcAgent`, there will be one thread blocking there waiting for the callback to finish. A better approach might be allowing something like the following, where there is no blocking:

```python
def udf():
  fr = torch.distributed.future_result()

  def callback(fm):
    fr.finish(2 * fm.get())  

  future_result = torch.distributed.future_result()
  fut = torch.distributed.rpc("worker2", torch.add, torch.ones(2, 2), 1)
  fut.then(callback)
```

pritamdamania87 commented:
In case of the UDF example mentioned above, doesn't the UDF anyways need to call `fut.wait()` in the end before returning? If so, how do we avoid blocking?

mrshenli commented:
@pritamdamania87 

> In case of the UDF example mentioned above, doesn't the UDF anyways need to call `fut.wait()` in the end before returning? If so, how do we avoid blocking?

It will not call `wait`, it will return immediately after `then`. The `RpcAgent` holds a `shared_ptr` to the `FutureMessage`, and the callback has already been registered in the `FutureMessage`, which will be fired when the response is ready. Do you see any issue with this API?

pritamdamania87 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -18,6 +18,13 @@ py::object to_py_obj(const Message& message) {
 
+FutureMessage::Callback wrap_callback(
+    FutureMessage& fut, const std::function<void(FutureMessage&)> py_cb) {
+  return [py_cb, &fut]() {
```
Why do we have this wrapper? We can just convert to py::object using `to_py_obj` and pass py::object to py_cb.

mrshenli replied:
Was initially thinking about avoiding conversion when unnecessary. I can make the change if that feels more nature, and we come back to address the overhead only if we do see that in the future.

pritamdamania87 commented on torch/csrc/distributed/rpc/FutureMessage.cpp:
```diff
@@ -52,7 +51,7 @@ void FutureMessage::fireCallbacks() {
   // Once completed_ is set to true, no one can add new callback to the list.
   for (auto& callback : callbacks) {
-    callback(message_);
+    callback();
```
When we write something like this: `FutureMessage.then(cb1).then(cb2)`, from a user point of view cb1 should act on the result of FutureMessage and cb2 should act on the result of cb1. In the current version, cb1 and cb2 both act on the result of FutureMessage. Also, FutureMessage.then(cb1) returns the same FutureMessage back. This is confusing for the user, since they might expect that the return value of `FutureMessage.then(cb1)` is a future that will be ready once cb1 is done.

I think the right way to do this is to have FutureMessage.then(cb1) return a new future with one callback. This new future waits for the original future to complete and then runs a single callback. Now, when we add .then(cb2), we create a new future which waits for the future from `FutureMessage.then(cb1)` to complete and runs on callback cb2. I feel this might be more intuitive for users.

mrshenli replied:
Make senses! I think it is necessary to have multiple Future object and support `then().then()` for asynchronous UDF. But as that is a non-goal for this PR, let me just change the API to `add_callback()` and disallow chaining, which should be enough to avoid confusion.

mrshenli commented on test/test_rpc.py:
```diff
@@ -110,5 +110,53 @@ def test_join_rpc(self):
         # it's safe to call join_rpc() multiple times
         dist.join_rpc()
 
+    @_wrap_with_rpc
```
Yes, I will address this in a separate PR after #23569 lands.

mrshenli commented on torch/csrc/distributed/rpc/FutureMessage.cpp:
```diff
@@ -40,8 +37,10 @@ bool FutureMessage::completed() const {
+    // invoking this callback in place does not violate the order.
     lock.unlock();
-    callback(message_);
+    callback();
```
It is not necessary, as we capture that in the lambda function. I am using this API here to stay consistent with `ivalue::Future`, as we are going to merge `FutureMessage` into `ivalue::Future` soon.

mrshenli commented on torch/csrc/distributed/rpc/FutureMessage.cpp:
```diff
@@ -52,7 +51,7 @@ void FutureMessage::fireCallbacks() {
   // Once completed_ is set to true, no one can add new callback to the list.
   for (auto& callback : callbacks) {
-    callback(message_);
+    callback();
```
ditto

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -41,7 +42,26 @@ PyObject* rpc_init(PyObject* /* unused */) {
           },
-          py::call_guard<py::gil_scoped_release>());
+          py::call_guard<py::gil_scoped_release>())
+      .def("get",
```
Good point on mentioning Java's and Folly's API. This was originally added to address [this comment](https://github.com/pytorch/pytorch/pull/23228#discussion_r309426549) in #23228. But I agree with you that staying consistent with Java and Folly might help to avoid surprising users. I will make the change.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -41,7 +42,26 @@ PyObject* rpc_init(PyObject* /* unused */) {
+          },
+          py::call_guard<py::gil_scoped_release>())
+      .def("then",
+          // Python Callback taking a FutureMessage& instead of py::object of
```
One case would be `FutureMessage.then(log_finish_time)`. Or more generally, there could be some scenarios using RPC to synchronize distributed execution by, e.g., modifying some remote global state. This would be especially useful if we do remove APIs like sync/join.

pritamdamania87 replied:
Could we have two variations of `.then` to address this? One in which the callback receives a `py::object` as input since the user needs the result and one which doesn't pass any parameters to the callback. We could call them `then` and `then_without_result`. I just feel that having the callback receive a Future isn't the right API since the user might think that this Future might not yet be ready (whereas its always ready). Although, I feel that for the first version we should have callback receive a py::object and if a user has needs where they don't need the result and the conversion from Message to py::object is an overhead, we could then introduce the `then_without_result` API.

mrshenli committed:
Update on "Support Callbacks on Asynchronous RPC"


RPC applications can now register callbacks to the FutureMessage
object returned by the asynchronous RPC API. Multiple callbacks can be
chained together by calling `FutureMessage.then(cb1).then(cb2)`. Each
callback should take one parameter, which is the reference to the
`FutureMessage` object. It is up to the callback whether it needs to
retrieve the return value or disregard.

Differential Revision: [D16743652](https://our.internmc.facebook.com/intern/diff/D16743652/)

pritamdamania87 commented:
> It will not call wait, it will return immediately after then. The RpcAgent holds a shared_ptr to the FutureMessage, and the callback has already been registered in the FutureMessage, which will be fired when the response is ready. Do you see any issue with this API?

If I call the udf with a synchronous rpc as follows from a different worker: `result = torch.distributed.rpc('worker1', 'udf');`. Is the expectation that once that RPC returns the `torch.distributed.rpc("worker2", torch.add, torch.ones(2, 2), 1)` operation in the UDF would have completed executing? IIUC, in the current implementation there is no guarantee that it would have completed executing. Its more like the UDF() kicked off some async work and returned to the caller. But the caller themselves have no way of knowing when this async work would be completed.

zhaojuanmao commented:
> > It will not call wait, it will return immediately after then. The RpcAgent holds a shared_ptr to the FutureMessage, and the callback has already been registered in the FutureMessage, which will be fired when the response is ready. Do you see any issue with this API?
> 
> If I call the udf with a synchronous rpc as follows from a different worker: `result = torch.distributed.rpc('worker1', 'udf');`. Is the expectation that once that RPC returns the `torch.distributed.rpc("worker2", torch.add, torch.ones(2, 2), 1)` operation in the UDF would have completed executing? IIUC, in the current implementation there is no guarantee that it would have completed executing. Its more like the UDF() kicked off some async work and returned to the caller. But the caller themselves have no way of knowing when this async work would be completed.

@pritamdamania yes, if UDF() kicked off async work, current rpc will immediately return without knowing async calls are completed. we need to address that, we can discuss this on next week's async python UDF meeting

mrshenli commented:
superseded by #37311
